### ISSUE TITLE
Potential false positive SC2320

### ISSUE THREAD

**Issue by Martiix** [koalaman/shellcheck#2655]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2320
- My shellcheck version (`shellcheck --version` or "online"): 0.9.0
- [ X] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [ X] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [ X] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [ X] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/bash
echo "some print that can fail" || exit $?

```

#### Here's what shellcheck currently says:
^-- SC2320 (warning): This $? refers to echo/printf, not a previous command. Assign to variable to avoid it being overwritten.

#### Here's what I wanted or expected to see:
I might be wrong, but I would assume that this shouldn't trigger a warning as the `$?` isn't part of the print command itself, and is expected to get the return code from the print statement. The || exit is meant to trigger if something goes wrong on the same line, therefore the return code in $? refers to the correct line. 

I use this format in almost all of my shell scripts, I don't want to ignore the warning as it should trigger in cases like the one mentioned in the wiki, so I'm wondering if this is expected behavior or if this is a bug that should be fixed so I can know if I need to change all my shell scripts or if this is accepted behavior in a later release. I haven't seen this warning before yesterday where I updated from shellcheck 0.8.0 to 0.9.0

**Comment by TinCanTech** [comment]
In your example, `exit` will inherit the value of `$?`, therefore, you can simply use `|| exit` to satisfy shellcheck.

**Comment by Martiix** [comment]
does the same go for 
`source "some/path/to/some/script.sh" || exit $?`
cause I still get the same complaints on this line

**Comment by TinCanTech** [comment]
The _same goes for_ your second question, too.

I _believe_ what `shellcheck` is trying to say is:
- Don't use `$?` directly, instead, assign it to a variable and use that.

I _believe_ this is due to a _common_ mistake, whereby, people use `$?` **after** it has been changed by a subsequent command.

This is why `shellcheck` gives a warning for it.

In your example, I would agree that this is probably a false-positive. However, @koalaman scans many thousands of scripts to get some idea of how common mistakes like this are, so he has a much better idea than I do.
